File: app/routers/sites.py
from __future__ import annotations

from fastapi import APIRouter, Form, Request
from fastapi.responses import PlainTextResponse, RedirectResponse
from sqlalchemy import select

from app.deps import CurrentUser, CurrentWorkspace, DbSession, flash, get_site_or_404
from app.google.errors import GoogleApiError
from app.google.indexnow import key_file_url
from app.models import IndexJob, PageUrl, Sitemap
from app.security import generate_indexnow_key
from app.services import quota, sitemaps as sitemap_service, sites as site_service, tasks
from app.services.scheduler import is_running, run_in_background
from app.services.stats import workspace_indexing_history
from app.services.urls import site_url_stats
from app.templating import render

router = APIRouter(prefix="/sites", tags=["sites"])


def _back(site_id: int, tab: str = "") -> RedirectResponse:
    suffix = f"?tab={tab}" if tab else ""
    return RedirectResponse(f"/sites/{site_id}{suffix}", status_code=303)


@router.get("")
def sites_page(request: Request, user: CurrentUser, workspace: CurrentWorkspace, db: DbSession):
    sites = site_service.list_sites(db, workspace)
    rows = []
    for site in sites:
        stats = site_url_stats(db, site.id)
        rows.append(
            {
                "site": site,
                "stats": stats,
                "sitemaps": db.scalar(
                    select(Sitemap).where(Sitemap.site_id == site.id).limit(1)
                )
                is not None,
                "busy": is_running(f"site:{site.id}"),
            }
        )
    return render(
        request,
        "sites.html",
        {
            "user": user,
            "workspace": workspace,
            "rows": rows,
            "active_page": "sites",
        },
    )


@router.post("/import")
def import_sites(request: Request, user: CurrentUser, workspace: CurrentWorkspace, db: DbSession):
    try:
        result = site_service.import_sites_from_gsc(db, user, workspace)
    except GoogleApiError as exc:
        flash(request, f"Nie udalo sie pobrac stron z Search Console: {exc}", "error")
        return RedirectResponse("/sites", status_code=303)

    flash(
        request,
        f"Import zakonczony: {result['created']} nowych, {result['updated']} zaktualizowanych"
        + (f", {result['skipped']} pominietych (brak weryfikacji)." if result["skipped"] else "."),
        "success",
    )
    return RedirectResponse("/sites", status_code=303)


@router.post("/add")
def add_site(
    request: Request,
    workspace: CurrentWorkspace,
    db: DbSession,
    property_url: str = Form(...),
):
    if not property_url.strip():
        flash(request, "Podaj adres strony.", "error")
        return RedirectResponse("/sites", status_code=303)

    site = site_service.create_site(db, workspace, property_url)
    flash(request, f"Dodano strone {site.display_name}.", "success")
    return _back(site.id)


@router.get("/{site_id}")
def site_detail(
    request: Request,
    site_id: int,
    user: CurrentUser,
    workspace: CurrentWorkspace,
    db: DbSession,
    tab: str = "overview",
    status: str = "",
    q: str = "",
    page: int = 1,
):
    site = get_site_or_404(db, workspace, site_id)
    per_page = 50

    query = select(PageUrl).where(PageUrl.site_id == site.id)
    if status:
        query = query.where(PageUrl.index_status == status)
    if q:
        query = query.where(PageUrl.url.contains(q))

    total = len(db.scalars(query).all())
    urls = list(
        db.scalars(
            query.order_by(PageUrl.index_status, PageUrl.id.desc())
            .offset((max(page, 1) - 1) * per_page)
            .limit(per_page)
        )
    )

    return render(
        request,
        "site_detail.html",
        {
            "user": user,
            "workspace": workspace,
            "site": site,
            "stats": site_url_stats(db, site.id),
            "sitemaps": list(
                db.scalars(select(Sitemap).where(Sitemap.site_id == site.id).order_by(Sitemap.id))
            ),
            "urls": urls,
            "total_urls": total,
            "page": max(page, 1),
            "per_page": per_page,
            "pages": max(1, (total + per_page - 1) // per_page),
            "filter_status": status,
            "query": q,
            "tab": tab,
            "jobs": list(
                db.scalars(
                    select(IndexJob)
                    .where(IndexJob.site_id == site.id)
                    .order_by(IndexJob.created_at.desc())
                    .limit(30)
                )
            ),
            "history": workspace_indexing_history(db, workspace.id, 30),
            "quota_left": quota.remaining(db, workspace),
            "busy": is_running(f"site:{site.id}"),
            "indexnow_url": key_file_url(site.home_url, site.indexnow_key or ""),
            "active_page": "sites",
        },
    )


@router.post("/{site_id}/settings")
def update_site_settings(
    request: Request,
    site_id: int,
    workspace: CurrentWorkspace,
    db: DbSession,
    display_name: str = Form(""),
    daily_limit: int = Form(50),
    priority: int = Form(0),
    auto_index: bool = Form(False),
    indexnow_enabled: bool = Form(False),
    is_active: bool = Form(False),
):
    site = get_site_or_404(db, workspace, site_id)
    if display_name.strip():
        site.display_name = display_name.strip()[:255]
    site.daily_limit = max(0, min(daily_limit, 10000))
    site.priority = max(0, min(priority, 100))
    site.auto_index = auto_index
    site.indexnow_enabled = indexnow_enabled
    site.is_active = is_active
    if indexnow_enabled and not site.indexnow_key:
        site.indexnow_key = generate_indexnow_key()
    db.commit()
    flash(request, "Ustawienia strony zapisane.", "success")
    return _back(site_id, "settings")


@router.post("/{site_id}/delete")
def delete_site(request: Request, site_id: int, workspace: CurrentWorkspace, db: DbSession):
    site = get_site_or_404(db, workspace, site_id)
    name = site.display_name
    db.delete(site)
    db.commit()
    flash(request, f"Usunieto strone {name} wraz z jej danymi.", "success")
    return RedirectResponse("/sites", status_code=303)


@router.post("/{site_id}/regenerate-key")
def regenerate_indexnow_key(
    request: Request, site_id: int, workspace: CurrentWorkspace, db: DbSession
):
    site = get_site_or_404(db, workspace, site_id)
    site.indexnow_key = generate_indexnow_key()
    db.commit()
    flash(request, "Wygenerowano nowy klucz IndexNow. Wgraj nowy plik na serwer.", "warning")
    return _back(site_id, "settings")


@router.get("/{site_id}/indexnow-key")
def download_indexnow_key(site_id: int, workspace: CurrentWorkspace, db: DbSession):
    site = get_site_or_404(db, workspace, site_id)
    key = site.indexnow_key or ""
    return PlainTextResponse(
        key,
        headers={"Content-Disposition": f'attachment; filename="{key}.txt"'},
    )


# ------------------------------------------------------------------ akcje


@router.post("/{site_id}/scan")
def scan_site(request: Request, site_id: int, workspace: CurrentWorkspace, db: DbSession):
    site = get_site_or_404(db, workspace, site_id)
    started = run_in_background(f"site:{site.id}", tasks.task_scan_sitemaps, site.id)
    flash(
        request,
        "Skanowanie sitemap uruchomione w tle." if started else "Zadanie dla tej strony juz trwa.",
        "success" if started else "warning",
    )
    return _back(site_id, "sitemaps")


@router.post("/{site_id}/inspect")
def inspect_site(
    request: Request,
    site_id: int,
    workspace: CurrentWorkspace,
    db: DbSession,
    limit: int = Form(50),
):
    site = get_site_or_404(db, workspace, site_id)
    started = run_in_background(f"site:{site.id}", tasks.task_inspect, site.id, limit)
    flash(
        request,
        f"Inspekcja {limit} URL-i uruchomiona w tle."
        if started
        else "Zadanie dla tej strony juz trwa.",
        "success" if started else "warning",
    )
    return _back(site_id, "urls")


@router.post("/{site_id}/run")
def run_pipeline(
    request: Request,
    site_id: int,
    workspace: CurrentWorkspace,
    db: DbSession,
    scan: bool = Form(True),
):
    site = get_site_or_404(db, workspace, site_id)
    if quota.remaining(db, workspace) <= 0:
        flash(request, "Dzienny limit zgloszen zostal wyczerpany.", "warning")
        return _back(site_id)

    started = run_in_background(f"site:{site.id}", tasks.task_run_pipeline, site.id, scan)
    flash(
        request,
        "Indeksowanie uruchomione: skan sitemap, inspekcja i zgloszenia."
        if started
        else "Zadanie dla tej strony juz trwa.",
        "success" if started else "warning",
    )
    return _back(site_id)


@router.post("/run-all")
def run_all(request: Request, workspace: CurrentWorkspace, db: DbSession):
    started = run_in_background(
        f"workspace:{workspace.id}", tasks.task_run_all_sites, workspace.id
    )
    flash(
        request,
        "Uruchomiono indeksowanie wszystkich stron."
        if started
        else "Indeksowanie juz trwa.",
        "success" if started else "warning",
    )
    return RedirectResponse("/sites", status_code=303)


# --------------------------------------------------------------- sitemapy


@router.post("/{site_id}/sitemaps/add")
def add_sitemap(
    request: Request,
    site_id: int,
    workspace: CurrentWorkspace,
    db: DbSession,
    path: str = Form(...),
    submit_to_google: bool = Form(False),
):
    site = get_site_or_404(db, workspace, site_id)
    sitemap = sitemap_service.add_sitemap(db, site, path)
    flash(request, f"Dodano sitemape {sitemap.path}.", "success")
    if submit_to_google:
        return RedirectResponse(
            f"/sites/{site_id}/sitemaps/{sitemap.id}/submit", status_code=307
        )
    return _back(site_id, "sitemaps")


@router.post("/{site_id}/sitemaps/sync")
def sync_sitemaps(
    request: Request, site_id: int, user: CurrentUser, workspace: CurrentWorkspace, db: DbSession
):
    site = get_site_or_404(db, workspace, site_id)
    try:
        result = sitemap_service.sync_from_gsc(db, user, site)
        flash(
            request,
            f"Zsynchronizowano sitemapy z GSC: {result['created']} nowych, "
            f"{result['updated']} zaktualizowanych.",
            "success",
        )
    except GoogleApiError as exc:
        flash(request, f"Blad synchronizacji sitemap: {exc}", "error")
    return _back(site_id, "sitemaps")


@router.post("/{site_id}/sitemaps/discover")
def discover_sitemaps(request: Request, site_id: int, workspace: CurrentWorkspace, db: DbSession):
    site = get_site_or_404(db, workspace, site_id)
    found = sitemap_service.discover_sitemaps(db, site)
    flash(
        request,
        f"Znaleziono {len(found)} sitemap." if found else "Nie znaleziono zadnej sitemapy.",
        "success" if found else "warning",
    )
    return _back(site_id, "sitemaps")


@router.post("/{site_id}/sitemaps/{sitemap_id}/submit")
def submit_sitemap(
    request: Request,
    site_id: int,
    sitemap_id: int,
    user: CurrentUser,
    workspace: CurrentWorkspace,
    db: DbSession,
):
    site = get_site_or_404(db, workspace, site_id)
    sitemap = db.get(Sitemap, sitemap_id)
    if sitemap is None or sitemap.site_id != site.id:
        flash(request, "Nie znaleziono sitemapy.", "error")
        return _back(site_id, "sitemaps")

    job = sitemap_service.submit_to_google(db, user, site, sitemap)
    flash(request, job.message or "Zgloszono sitemape.", "success" if job.status == "SUCCESS" else "error")
    return _back(site_id, "sitemaps")


@router.post("/{site_id}/sitemaps/{sitemap_id}/scan")
def scan_single_sitemap(
    request: Request, site_id: int, sitemap_id: int, workspace: CurrentWorkspace, db: DbSession
):
    site = get_site_or_404(db, workspace, site_id)
    sitemap = db.get(Sitemap, sitemap_id)
    if sitemap is None or sitemap.site_id != site.id:
        flash(request, "Nie znaleziono sitemapy.", "error")
        return _back(site_id, "sitemaps")

    result = sitemap_service.scan_sitemap(db, site, sitemap)
    if result.get("error"):
        flash(request, f"Blad skanowania: {result['error']}", "error")
    else:
        flash(
            request,
            f"Zaimportowano {result['added']} nowych URL-i "
            f"({result['duplicates']} juz istnialo).",
            "success",
        )
    return _back(site_id, "sitemaps")


@router.post("/{site_id}/sitemaps/{sitemap_id}/delete")
def delete_sitemap(
    request: Request,
    site_id: int,
    sitemap_id: int,
    user: CurrentUser,
    workspace: CurrentWorkspace,
    db: DbSession,
    from_google: bool = Form(False),
):
    site = get_site_or_404(db, workspace, site_id)
    sitemap = db.get(Sitemap, sitemap_id)
    if sitemap is None or sitemap.site_id != site.id:
        flash(request, "Nie znaleziono sitemapy.", "error")
        return _back(site_id, "sitemaps")

    if from_google:
        job = sitemap_service.delete_from_google(db, user, site, sitemap)
        if job.status != "SUCCESS":
            flash(request, f"Google: {job.message}", "warning")

    db.delete(sitemap)
    db.commit()
    flash(request, "Sitemapa usunieta.", "success")
    return _back(site_id, "sitemaps")

File: src/routes/sites.ts
import { Router } from "express";
import { db } from "../db/index.js";
import { IndexStatus, JobStatus } from "../db/types.js";
import { GoogleApiError } from "../google/errors.js";
import { keyFileUrl } from "../google/indexnow.js";
import { generateIndexNowKey } from "../lib/crypto.js";
import { asyncHandler, flash, requireSite } from "../middleware/auth.js";
import * as quota from "../services/quota.js";
import { isRunning, runInBackground } from "../services/scheduler.js";
import * as sitemapService from "../services/sitemaps.js";
import * as siteService from "../services/sites.js";
import { workspaceIndexingHistory } from "../services/stats.js";
import * as tasks from "../services/tasks.js";
import { siteUrlStats } from "../services/urls.js";
import { baseContext } from "../templating.js";
import { panelAuth } from "./auth.js";

export const sitesRouter = Router();
sitesRouter.use(...panelAuth);

function back(siteId: number, tab = ""): string {
  return tab ? `/sites/${siteId}?tab=${tab}` : `/sites/${siteId}`;
}

function formBool(value: unknown): boolean {
  if (Array.isArray(value)) return value.some((v) => formBool(v));
  return value === "on" || value === "true" || value === "1" || value === true;
}

sitesRouter.get(
  "/",
  asyncHandler(async (req, res) => {
    const workspace = req.workspace!;
    const sites = await siteService.listSites(workspace.id);
    const rows = [];
    for (const site of sites) {
      const stats = await siteUrlStats(site.id);
      const hasSitemap = await db
        .selectFrom("sitemaps")
        .select("id")
        .where("site_id", "=", site.id)
        .limit(1)
        .executeTakeFirst();
      rows.push({
        site,
        stats,
        sitemaps: Boolean(hasSitemap),
        busy: isRunning(`site:${site.id}`),
      });
    }
    res.render(
      "sites.html",
      baseContext(req, { user: req.user, workspace, rows, active_page: "sites" }),
    );
  }),
);

sitesRouter.post(
  "/import",
  asyncHandler(async (req, res) => {
    try {
      const result = await siteService.importSitesFromGsc(req.user!.id, req.workspace!);
      flash(
        req,
        `Import zakonczony: ${result.created} nowych, ${result.updated} zaktualizowanych` +
          (result.skipped ? `, ${result.skipped} pominietych (brak weryfikacji).` : "."),
        "success",
      );
    } catch (error) {
      const message = error instanceof GoogleApiError ? error.toString() : String(error);
      flash(req, `Nie udalo sie pobrac stron z Search Console: ${message}`, "error");
    }
    res.redirect(303, "/sites");
  }),
);

sitesRouter.post(
  "/add",
  asyncHandler(async (req, res) => {
    const propertyUrl = String(req.body.property_url ?? "").trim();
    if (!propertyUrl) {
      flash(req, "Podaj adres strony.", "error");
      return res.redirect(303, "/sites");
    }
    const site = await siteService.createSite(req.workspace!.id, propertyUrl);
    flash(req, `Dodano strone ${site.display_name}.`, "success");
    res.redirect(303, back(site.id));
  }),
);

sitesRouter.post(
  "/run-all",
  asyncHandler(async (req, res) => {
    const workspace = req.workspace!;
    const started = runInBackground(`workspace:${workspace.id}`, () =>
      tasks.taskRunAllSites(workspace.id),
    );
    flash(
      req,
      started ? "Uruchomiono indeksowanie wszystkich stron." : "Indeksowanie juz trwa.",
      started ? "success" : "warning",
    );
    res.redirect(303, "/sites");
  }),
);

sitesRouter.get(
  "/:siteId",
  asyncHandler(async (req, res) => {
    const siteId = Number(req.params.siteId);
    const site = await requireSite(req, siteId);
    const perPage = 50;
    const tab = typeof req.query.tab === "string" ? req.query.tab : "overview";
    const status = typeof req.query.status === "string" ? req.query.status : "";
    const q = typeof req.query.q === "string" ? req.query.q : "";
    const page = Math.max(1, Number(req.query.page) || 1);

    let query = db.selectFrom("urls").selectAll().where("site_id", "=", site.id);
    if (status) query = query.where("index_status", "=", status as IndexStatus);
    if (q) query = query.where("url", "like", `%${q}%`);

    const totalRow = await query
      .clearSelect()
      .select((eb) => eb.fn.countAll<number>().as("total"))
      .executeTakeFirst();
    const total = Number(totalRow?.total ?? 0);

    const urls = await query
      .orderBy("index_status")
      .orderBy("id", "desc")
      .offset((page - 1) * perPage)
      .limit(perPage)
      .execute();

    const sitemaps = await sitemapService.listSitemaps(site.id);
    const jobs = await db
      .selectFrom("index_jobs")
      .selectAll()
      .where("site_id", "=", site.id)
      .orderBy("created_at", "desc")
      .limit(30)
      .execute();

    res.render(
      "site_detail.html",
      baseContext(req, {
        user: req.user,
        workspace: req.workspace,
        site,
        stats: await siteUrlStats(site.id),
        sitemaps,
        urls,
        total_urls: total,
        page,
        per_page: perPage,
        pages: Math.max(1, Math.ceil(total / perPage)),
        filter_status: status,
        query: q,
        tab,
        jobs,
        history: await workspaceIndexingHistory(req.workspace!.id, 30),
        quota_left: await quota.remaining(req.workspace!),
        busy: isRunning(`site:${site.id}`),
        indexnow_url: keyFileUrl(site.home_url, site.indexnow_key || ""),
        active_page: "sites",
      }),
    );
  }),
);

sitesRouter.post(
  "/:siteId/settings",
  asyncHandler(async (req, res) => {
    const site = await requireSite(req, Number(req.params.siteId));
    const displayName = String(req.body.display_name ?? "").trim();
    const dailyLimit = Math.max(0, Math.min(Number(req.body.daily_limit) || 50, 10_000));
    const priority = Math.max(0, Math.min(Number(req.body.priority) || 0, 100));
    const indexnowEnabled = formBool(req.body.indexnow_enabled);
    const isActive = formBool(req.body.is_active);

    await db
      .updateTable("sites")
      .set({
        ...(displayName ? { display_name: displayName.slice(0, 255) } : {}),
        daily_limit: dailyLimit,
        priority,
        // auto_index ma wlasny endpoint /toggle-auto — nie nadpisuj go tu
        indexnow_enabled: indexnowEnabled,
        is_active: isActive,
        ...(indexnowEnabled && !site.indexnow_key
          ? { indexnow_key: generateIndexNowKey() }
          : {}),
      })
      .where("id", "=", site.id)
      .execute();

    flash(req, "Ustawienia strony zapisane.", "success");
    res.redirect(303, back(site.id, "settings"));
  }),
);

sitesRouter.post(
  "/:siteId/toggle-auto",
  asyncHandler(async (req, res) => {
    const site = await requireSite(req, Number(req.params.siteId));
    // Stan z checkboxa po onchange (nie !db — unikamy race / zlej konwersji TINYINT).
    const enabled =
      "enabled" in (req.body as object)
        ? formBool(req.body.enabled)
        : !Boolean(site.auto_index);
    await db
      .updateTable("sites")
      .set({ auto_index: enabled })
      .where("id", "=", site.id)
      .execute();
    flash(
      req,
      enabled
        ? `Auto-indeks wlaczony dla ${site.display_name} (limit ${site.daily_limit}/dzien).`
        : `Auto-indeks wylaczony dla ${site.display_name}.`,
      "success",
    );
    const redirectTo = String(req.body.redirect || "/sites");
    res.redirect(303, redirectTo.startsWith("/") ? redirectTo : "/sites");
  }),
);

sitesRouter.post(
  "/:siteId/delete",
  asyncHandler(async (req, res) => {
    const site = await requireSite(req, Number(req.params.siteId));
    const name = site.display_name;
    await siteService.deleteSite(site.id, req.workspace!.id);
    flash(req, `Usunieto strone ${name} wraz z jej danymi.`, "success");
    res.redirect(303, "/sites");
  }),
);

sitesRouter.post(
  "/:siteId/regenerate-key",
  asyncHandler(async (req, res) => {
    const site = await requireSite(req, Number(req.params.siteId));
    await db
      .updateTable("sites")
      .set({ indexnow_key: generateIndexNowKey() })
      .where("id", "=", site.id)
      .execute();
    flash(req, "Wygenerowano nowy klucz IndexNow. Wgraj nowy plik na serwer.", "warning");
    res.redirect(303, back(site.id, "settings"));
  }),
);

sitesRouter.get(
  "/:siteId/indexnow-key",
  asyncHandler(async (req, res) => {
    const site = await requireSite(req, Number(req.params.siteId));
    const key = site.indexnow_key || "";
    res.setHeader("Content-Disposition", `attachment; filename="${key}.txt"`);
    res.type("text/plain").send(key);
  }),
);

sitesRouter.post(
  "/:siteId/scan",
  asyncHandler(async (req, res) => {
    const site = await requireSite(req, Number(req.params.siteId));
    const started = runInBackground(`site:${site.id}`, () => tasks.taskScanSitemaps(site.id));
    flash(
      req,
      started ? "Skanowanie sitemap uruchomione w tle." : "Zadanie dla tej strony juz trwa.",
      started ? "success" : "warning",
    );
    res.redirect(303, back(site.id, "sitemaps"));
  }),
);

sitesRouter.post(
  "/:siteId/inspect",
  asyncHandler(async (req, res) => {
    const site = await requireSite(req, Number(req.params.siteId));
    const limit = Math.max(1, Number(req.body.limit) || 50);
    const started = runInBackground(`site:${site.id}`, () => tasks.taskInspect(site.id, limit));
    flash(
      req,
      started
        ? `Inspekcja ${limit} URL-i uruchomiona w tle.`
        : "Zadanie dla tej strony juz trwa.",
      started ? "success" : "warning",
    );
    res.redirect(303, back(site.id, "urls"));
  }),
);

sitesRouter.post(
  "/:siteId/run",
  asyncHandler(async (req, res) => {
    const site = await requireSite(req, Number(req.params.siteId));
    if ((await quota.remaining(req.workspace!)) <= 0) {
      flash(req, "Dzienny limit zgloszen zostal wyczerpany.", "warning");
      return res.redirect(303, back(site.id));
    }
    const scan = formBool(req.body.scan) || req.body.scan === undefined;
    const started = runInBackground(`site:${site.id}`, () => tasks.taskRunPipeline(site.id, scan));
    flash(
      req,
      started
        ? "Indeksowanie uruchomione: skan sitemap, inspekcja i zgloszenia."
        : "Zadanie dla tej strony juz trwa.",
      started ? "success" : "warning",
    );
    res.redirect(303, back(site.id));
  }),
);

sitesRouter.post(
  "/:siteId/sitemaps/add",
  asyncHandler(async (req, res) => {
    const site = await requireSite(req, Number(req.params.siteId));
    const path = String(req.body.path ?? "").trim();
    const sitemap = await sitemapService.addSitemap(site, path);
    flash(req, `Dodano sitemape ${sitemap.path}.`, "success");
    if (formBool(req.body.submit_to_google)) {
      const job = await sitemapService.submitToGoogle(req.user!.id, site, sitemap);
      flash(
        req,
        job.message || "Zgloszono sitemape.",
        job.status === JobStatus.SUCCESS ? "success" : "error",
      );
    }
    res.redirect(303, back(site.id, "sitemaps"));
  }),
);

sitesRouter.post(
  "/:siteId/sitemaps/sync",
  asyncHandler(async (req, res) => {
    const site = await requireSite(req, Number(req.params.siteId));
    try {
      const result = await sitemapService.syncFromGsc(req.user!.id, site);
      flash(
        req,
        `Zsynchronizowano sitemapy z GSC: ${result.created} nowych, ${result.updated} zaktualizowanych.`,
        "success",
      );
    } catch (error) {
      const message = error instanceof GoogleApiError ? error.toString() : String(error);
      flash(req, `Blad synchronizacji sitemap: ${message}`, "error");
    }
    res.redirect(303, back(site.id, "sitemaps"));
  }),
);

sitesRouter.post(
  "/:siteId/sitemaps/discover",
  asyncHandler(async (req, res) => {
    const site = await requireSite(req, Number(req.params.siteId));
    const found = await sitemapService.discoverSitemaps(site);
    flash(
      req,
      found.length ? `Znaleziono ${found.length} sitemap.` : "Nie znaleziono zadnej sitemapy.",
      found.length ? "success" : "warning",
    );
    res.redirect(303, back(site.id, "sitemaps"));
  }),
);

sitesRouter.post(
  "/:siteId/sitemaps/:sitemapId/submit",
  asyncHandler(async (req, res) => {
    const site = await requireSite(req, Number(req.params.siteId));
    const sitemap = await sitemapService.getSitemap(Number(req.params.sitemapId), site.id);
    if (!sitemap) {
      flash(req, "Nie znaleziono sitemapy.", "error");
      return res.redirect(303, back(site.id, "sitemaps"));
    }
    const job = await sitemapService.submitToGoogle(req.user!.id, site, sitemap);
    flash(
      req,
      job.message || "Zgloszono sitemape.",
      job.status === JobStatus.SUCCESS ? "success" : "error",
    );
    res.redirect(303, back(site.id, "sitemaps"));
  }),
);

sitesRouter.post(
  "/:siteId/sitemaps/:sitemapId/scan",
  asyncHandler(async (req, res) => {
    const site = await requireSite(req, Number(req.params.siteId));
    const sitemap = await sitemapService.getSitemap(Number(req.params.sitemapId), site.id);
    if (!sitemap) {
      flash(req, "Nie znaleziono sitemapy.", "error");
      return res.redirect(303, back(site.id, "sitemaps"));
    }
    const result = await sitemapService.scanSitemap(site, sitemap);
    if (result.error) flash(req, `Blad skanowania: ${result.error}`, "error");
    else {
      flash(
        req,
        `Zaimportowano ${result.added} nowych URL-i (${result.duplicates} juz istnialo).`,
        "success",
      );
    }
    res.redirect(303, back(site.id, "sitemaps"));
  }),
);

sitesRouter.post(
  "/:siteId/sitemaps/:sitemapId/delete",
  asyncHandler(async (req, res) => {
    const site = await requireSite(req, Number(req.params.siteId));
    const sitemap = await sitemapService.getSitemap(Number(req.params.sitemapId), site.id);
    if (!sitemap) {
      flash(req, "Nie znaleziono sitemapy.", "error");
      return res.redirect(303, back(site.id, "sitemaps"));
    }
    if (formBool(req.body.from_google)) {
      const job = await sitemapService.deleteFromGoogle(req.user!.id, site, sitemap);
      if (job.status !== JobStatus.SUCCESS) flash(req, `Google: ${job.message}`, "warning");
    }
    await sitemapService.deleteSitemap(sitemap.id, site.id);
    flash(req, "Sitemapa usunieta.", "success");
    res.redirect(303, back(site.id, "sitemaps"));
  }),
);
